### ISSUE TITLE
Fixpoint not reached in some incremental runs on chrony

### ISSUE THREAD

**Issue by jerhard** [goblint/analyzer#1091]
Using the current `master`, 35 runs of the incremental analysis do not reach the fixpoint on chrony. This occurred with the regular incremental configuration. In all of the 35 runs, the unreached fixpoint is detected at the same code location within `chrony`.
In the experiments, it does not manifest in the runs where the incremental postsolver is active (with reluctant incremental analysis on or off). 

<details>
<summary> Difference between rhs and computed solution.</summary>

```
Difference: Map: Reversed (top or Set (address (lval (offset (IntDomLifter(intdomtuple)))) * booleans)) * bottom or map:({}, mapping {
                                                                                                                                }) =
 Map: messages =
 trivial arrays: Array: mapping {
                          addr_type ->   Uninitialized
                          data ->   {}
                          descriptor ->   Uninitialized
                          if_index ->   Uninitialized
                          length ->   Uninitialized
                          local_addr ->
                            (If you see this, you are special!, Unknown)
                          remote_addr ->
                            (If you see this, you are special!, Unknown)
                          timestamp ->
                            mapping {
                              hw ->   mapping {
                                        tv_nsec ->   Uninitialized
                                        tv_sec ->   Uninitialized
                                      }
                              if_index ->   Uninitialized
                              kernel ->   mapping {
                                            tv_nsec ->   Uninitialized
                                            tv_sec ->   Uninitialized
                                          }
                              l2_length ->   Uninitialized
                              tx_flags ->   Uninitialized
                            }
                        } not leq Array: mapping {
                                             addr_type ->   Uninitialized
                                             data ->   {}
                                             descriptor ->   Uninitialized
                                             if_index ->   Uninitialized
                                             length ->   Uninitialized
                                             local_addr ->
                                               (If you see this, you are special!, Uninitialized)
                                             remote_addr ->
                                               (If you see this, you are special!, Uninitialized)
                                             timestamp ->
                                               mapping {
                                                 hw ->   mapping {
                                                           tv_nsec ->
                                                             Uninitialized
                                                           tv_sec ->
                                                             Uninitialized
                                                         }
                                                 if_index ->   Uninitialized
                                                 kernel ->   mapping {
                                                               tv_nsec ->
                                                                 Uninitialized
                                                               tv_sec ->
                                                                 Uninitialized
                                                             }
                                                 l2_length ->   Uninitialized
                                                 tx_flags ->   Uninitialized
                                               }
                                           }
```
</details>

**Comment by jerhard** [comment]
The fixpoint is not reached at a call:
```
Fixpoint not reached at L:node 47288 "received = SCK_ReceiveMessages(sock_fd,
                                                                     messages,
                                                                     4, flags);" on chrony/ntp_io.c:426:3-426:62
```
The line number `chrony/ntp_io.c:426:3-426:62` is reported on [this small commit](https://github.com/mlichvar/chrony/commit/934b8712a57e324581a15ddb4f6c84cd67a5673e), where the fixpoint issue also arises.

**Comment by jerhard** [comment]
@stilscher pointed out that this may have to do with anonymous types in C. Indeed, it turns out the type of `local_addr` is an anonymous union or anonymous struct (in the commits I've checked), see here: https://github.com/mlichvar/chrony/blob/934b8712a57e324581a15ddb4f6c84cd67a5673e/socket.h#L64-L66

Previously, there were similar fixpoint issues that were caused by different names and/or `ckey`s assigned to the anonymous types by CIL. This may be a similar issue.

**Comment by jerhard** [comment]
I had a look at the  [previously mentioned commit](https://github.com/mlichvar/chrony/commit/26a1ed8bc349953814ec2df1ee27dcd4106bec72) where only a function was added. For the parent commit of that commit, at the node, where the incremental analysis of the child does not reach the fixpoint, the solver computes `local_addr -> (If you see this, you are special!, Uninitialized)` as does the right-hand side in the postsolving. 

From this, I would conclude for the child commit, that the right-hand-side that is not subsumed, that yields `local_addr ->  (If you see this, you are special!, Unknown)`, may be the problem. For the parent commit the computed solution subsumes the right-hand-side; and there are no relevant code changes that should change that fact.

<details>
<summary>Computed value in non-incremental run</summary>

```
{([Unit:(), Reversed (top or Set (varinfo)):{}, top or Set (Set (exp)):{},
    Reversed (top or Set (exp)):{}, Unit:(),
    Reversed (top or Set (address (lval (offset (IntDomLifter(intdomtuple)))) * booleans)) * bottom or map:({}, mapping {
                                                                                                                  }),
    Unit:(), Unit:(), top or Set (variables):{line}, booleans:False,
    MT mode:Multithreaded (main),
    lifted varinfo * node * chain option * Thread * lifted created (once) * created (multiple times) and Unit:(unknown node, main, ()),
    value domain * array partitioning deps * Vars with Weak Update * Unit:(mapping {
                                                                             Local {
                                                                               messages ->
                                                                                 (Array: mapping {
                                                                                           addr_type ->
                                                                                             Uninitialized
                                                                                           data ->
                                                                                             {}
                                                                                           descriptor ->
                                                                                             Uninitialized
                                                                                           if_index ->
                                                                                             Uninitialized
                                                                                           length ->
                                                                                             Uninitialized
                                                                                           local_addr ->
                                                                                             (If you see this, you are special!, Uninitialized)
                                                                                           remote_addr ->
                                                                                             (If you see this, you are special!, Uninitialized)
                                                                                           timestamp ->
                                                                                             mapping {
                                                                                               hw ->
                                                                                                 mapping {
                                                                                                   tv_nsec ->
                                                                                                     Uninitialized
                                                                                                   tv_sec ->
                                                                                                     Uninitialized
                                                                                                 }
                                                                                               if_index ->
                                                                                                 Uninitialized
                                                                                               kernel ->
                                                                                                 mapping {
                                                                                                   tv_nsec ->
                                                                                                     Uninitialized
                                                                                                   tv_sec ->
                                                                                                     Uninitialized
                                                                                                 }
                                                                                               l2_length ->
                                                                                                 Uninitialized
                                                                                               tx_flags ->
                                                                                                 Uninitialized
                                                                                             }
                                                                                         }, (4,[4,4]))
                                                                               i ->
                                                                                 (Unknown int([-31,31]),[-2147483648,2147483647])
                                                                               received ->
                                                                                 (Unknown int([-31,31]),[-2147483648,2147483647])
                                                                               flags ->
                                                                                 (Unknown int([0,1]),[0,1])
                                                                             }
                                                                             Parameter {
                                                                               sock_fd ->
                                                                                 (Unknown int([-31,31]),[-2147483648,2147483647])
                                                                               event ->
```
</summary>

**Comment by jerhard** [comment]
The first commit of Goblint in the recent past where `Fixpoint` appears for the output logs on a select subset of `chrony`-commits is this [commit](8d22c4572ad455c86673bd4f51ca50463f1e8212). 

I still have to have a look whether there are other issues at the previous commit that may mask the issue in the previous commit (e.g. if more runs were failing due to other reasons).

**Comment by jerhard** [comment]
Undoing the changes to the `relift` functions as in `d34cd61ef00534cdd552e7a7ca9f83c544d5ce84` indeed leads to the Fixpoint warnings going away on the `chrony` commits tested. 
Since the the `relift` functions were looking meaningful, it may be that reverting back justs hides the issue somehow. It may also be that the implementation of the `relift` in some of these places may have lead to inconsistencies compared to modules where the relifting is not done.

**Comment by michael-schwarz** [comment]
This morning @jerhard and I were able to narrow it down: It is the addition of relifting to the CPA, that causes the analysis to start failing. Curiously, CPA is `HashCached`, so that may have something to do with it.

**Comment by jerhard** [comment]
I removed the `HashCached` Lifter for the `CPA`, and the fixpoint issue remains.

**Comment by jerhard** [comment]
When deactivating the `relift` of arrays in `ValueDomain.relift` on the current `master`-branch, no fixpoints issues are reported on the tested `chrony` commits.

**Comment by michael-schwarz** [comment]
That is very strange: You said you are not using partitioned arrays, right?
